readme.md
This at the moment it is little more that a minimal front end to
libz80, while there are other z80 IDE's I'm developing this for
intellectual exercise. Its main feature will be to have the ability
to use devices provided by plugins.

The idea is that each plugin *instance* will have its own configuration that
will define what memory area (and/or IO ports) its responsible for 

The user will then be able to "build" their own Z80 system by including
and defining which devices (plugins) and how they are attached.

requires gtk3 pkg-config z88dk libz80

http://nightly.z88dk.org/z88dk-latest.tgz
https://github.com/ggambetta/libz80

extract z88dk to this directory
clone libz80 to this directory

![example virtual machine](runz80.png)

./runz80 machine1.xml

will load a machine that will allow you to run hello-input.bin

You should see Hello World! scrolling down the screen if you change the
value of the 8 bit switches, the hex value is displayed in the top
left of the display

example machine config

The following config is using two plugins, simpleIn is used once, and simpleOut is used twice so the
machine has three devices...

the user specifies the first port for the device (the plugin knows how many ports it needs!)

    <machine name="machine1">
        <plugin
            libName="simpleIn"
            label="input switches"
            portStart="0x10" />

        <plugin
            libName="simpleOut"
            label="input value"
            portStart="0x12" />

        <plugin
            libName="simpleOut"
            label="inverted value"
            portStart="0x14" />
    </machine>

having a "machine" like this you might use it with the following z80 code

        org 0
        
    loop:
        in  a,(0x10)    # simpleIn plugin port
        out (0x12),a    # simpleOut port
        cp  0x80
        jr  z, done
        cpl             # invert
        out (0x14),a    # the other simpleOut port
        jr  loop
    done:
        halt


in addition to the simpleIn and simpleOut plugin there is also kBuffer which
provides 1KB of bit pattern monochrome display, giving a luxurious 128x64x1
stunning display ability. In addition to the 1kb of memory mapped interface
(for the display bitmap) there is also a single port used, any value written
to this port will result in the actual display being updated from the memory.
This allows the whole image to be rendered into the memory but only refreshed
once its ready to be displayed.



z80asm-1.8/ChangeLog
2009-04-11  Bas Wijnen  <wijnen@debian.org>

	* z80asm.c: Fixed bug that defw/dw didn't accept more than one
	argument.

2009-01-18  Bas Wijnen  <wijnen@debian.org>

	* expressions.c: Fix incorrect valid declaration of expressions when
	only the last label is valid.  Thanks to Tomaž Šolc for reporting.

2007-09-09  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c: Wrap program counter consistently.
	* z80asm.c, z80asm.h: Fix error reporting in references.
	* all files: Update license to GPL version 3 or later.
	* Makefile: Improve make dist, and make it work with git.

2007-06-14  Bas Wijnen  <shevek@fmf.nl>

	* VERSION: Updated to 1.5
	* z80asm.c, z80asm.h: Fix handling of $ in stored expressions (thanks
	to Tomaz Solc for reporting and analysing this.
	* Makefile: Remove executable if tests fail.
	* tests/pass.asm, tests/pass.correct-bin: Add some tests.
	* z80asm.c: Make unresolvable references a normal error (instead of an
	internal assembler problem).

2007-06-05  Bas Wijnen  <shevek@fmf.nl>

	* VERSION: Updated to 1.4
	* z80asm.c: Fix bit/set/res instructions.

2007-05-24  Bas Wijnen  <wijnen@debian.org>

	* VERSION: Updated to 1.3.
	* tests, tests/pass.asm, tests/pass.correct-bin,
	tests/pass.correct-err, tests/Makefile: Add test suite to spot
	regressions.
	* expressions.c, z80asm.h: New files, split from z80asm.c.
	* Makefile: Updated.
	* expressions.c, z80asm.c: Warn for value truncation, warn for
	expressions fully in parenthesis, improve skipword so it doesn't break
	on comments with parentheses, improve expression parsing so
	"ld a, (1) + 1" doesn't result in an error.
	* examples/macro.asm: Fixed to make it usable as a rom.
	* z80asm.1: Updated.

2005-11-30  Jan Wilmans  <jw@dds.nl>

	* z80asm.c: Added unoffical syntax of ADD A,r as ADD r
	* z80asm.c: Added unoffical syntax of SUB r as SUB A,r

2005-11-30  Jan Wilmans  <jw@dds.nl>

	* z80asm.c: Added new seek command

2005-11-29  Jan Wilmans  <jw@dds.nl>

	* z80asm.c: Added two extra verbosity levels.

2005-11-13  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c: Run indent.

2005-10-14  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c (struct reference, compute_ref, new_reference): Removed
	copy of stack frame from struct reference.

2005-09-14  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c (open_include_file, assemble): Open incbin'd files as
	binary.

2005-06-16  Bas Wijnen  <shevek@fmf.nl>

	* Makefile: Fix incorrect version.
	* z80asm.c (enum mnemonic, mnemonics, assemble), examples/hello.asm:
	Rename bininclude to incbin.
	* z80asm.c (readcommand, assemble), examples/macro.asm: Allow labels
	before endm and endif.

2005-05-19  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c (struct reference, printerr, compute_ref, new_reference,
	assemble): Added stack frame to reference.
	* z80asm.c (labelfilename, parse_commandline): Remove output files
	when compilation fails.
	* z80asm.c (use_force, parse_commandline, assemble, main): Implement
	--force.
	* z80asm.c (parse_commandline): Make a.bin the default target.
	* z80asm.c (check_label, rd_label, rd_value, rd_expr,
	get_include_name, assemble): Give errors about junk at end of
	line or expression.
	* z80asm.c (rd_value): Support double quotes for character constants.
	* z80asm.c (rd_expr_equal): Support = for equality.
	* z80asm.1: Updated manpage.

2005-05-18  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c (enum mnemonic, enum reftype, struct reference, struct
	label, struct stack, struct macro_arg, struct macro_line, struct
	macro, mnemonics, firstmacro, define_macro, sp, stack, readlabel,
	rd_expr, compute_ref, check_label, rd_label, rd_value, rd_factor,
	rd_term, rd_expr_shift, rd_expr_unequal, rd_expr_equal, rd_expr_and,
	rd_expr_xor, rd_expr_or, new_reference, wrt_ref, read_line,
	get_macro_args, assemble): Added macro support.
	* z80asm.c (rd_value): Added several radix notations.
	* examples/macro.asm: New file.

2005-05-17  Bas Wijnen  <shevek@fmf.nl>

	* Makefile (clean): Clean examples and headers.
	* Makefile (dist): Make versioned tarballs.
	* z80asm.c (buffer, try_use_real_file, flush_to_real_file,
	parse_commandline, assemble): Generalise writing to non-seekable
	files.
	* z80asm.c (reallistfile, parse_commandline, assemble): Use it for
	list file.
	* z80asm.c (read_line, assemble): Remove arbitrary line length limit.

2005-05-13  Bas Wijnen  <shevek@fmf.nl>

	* examples/Makefile, examples/hello.asm, headers/msx-bios.asm,
	headers/msx2-bios.asm, headers/msx2+-bios.asm,
	headers/msxturbor-bios.asm: New files.
	* z80asm.c (struct includedir, firstincludedir, open_include_file,
	add_include, parse_commandline, assemble, main): Use include path.
	* z80asm (parse_commandline): Make stderr default for list and label
	files.
	* z80asm.c (enum mnemonic, mnemonics, get_include_name, assemble): Add
	bininclude directive.
	* z80asm.c (printerr, assemble): Make error output parsable.
	* z80asm.1: Updated.

2005-05-11  Bas Wijnen  <shevek@fmf.nl>

	* Makefile (top level, dist): Use versions instead of dates.
	* VERSION: New file.
	* BUGS, NEWS, README: Removed.

2005-05-10  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.1: New file.
	* z80asm.c (enum mnemonic, mnemonic, assemble): Support END directive.
	* z80asm.c (compare, indx, readlabel): Use strncasecmp instead of
	compare.
	* z80asm.c (assemble): Fix bug with strings in DEFB.
	* z80asm.c (assemble): Allow backslash-escapes in DEFB strings.
	* z80asm.c (rd_character): Allow octal escapes in strings.

2005-03-11  Bas Wijnen  <shevek@fmf.nl>

	* z80asm.c (rd_r): Fixed index prefix.
	* z80asm.c (assemble): Fixed register count errors.

2004-10-01  Bas Wijnen  <b.wijnen@phys.rug.nl>

	* z80asm (rd_comma): Fixed possible buffer overflow.
	* z80asm (parse_commandline): Improved --help output.
	* z80asm (new_reference): Removed useless ++.
	* z80asm (wrt_ref, assemble): Improved list output for ds.

2004-10-01  H. Peter Anvin  <hpa@users.sourceforge.net>

	* z80asm.c (enum mnemonic, mnemonics, assemble): Added dm/defm as an
	alias for db/defb.

main.c


#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>


#include "libz80/z80.h"
#include "main.h"

#include <gmodule.h>
#include "plugin.h"

#include <gtk/gtk.h>
#include <expat.h>

GtkWidget *mainwin, *runButton;
GtkWidget *AFtext,*BCtext,*DEtext,*HLtext;
GtkWidget *IXtext,*IYtext,*SPtext,*PCtext;
GtkWidget *addressInput,*memCont1,*memCont2;
GtkWidget *pluginPanel;

/* 64Kb of RAM */
static byte memory[ 0x10000 ];
static Z80Context context;
char tmpstr[80];
gboolean running = FALSE;

// list of plugin instances
GList *plugins;
guint idle_id;      // thread idle id

gboolean inhibitFocus=FALSE;

gboolean on_mainWindow_focus(GtkWidget *widget, GdkEvent  *event, gpointer   user_data)
{
    /* lot easier now all in one window...*/
    return TRUE;
}



void close_app(GtkWidget* widget,gpointer user_data)
{
    gtk_main_quit();
}

void on_reset(GtkWidget* widget,gpointer user_data)
{
    Z80RESET(&context);
    dump_z80_state();
}

void on_addressChange(GtkWidget* widget,gpointer user_data)
{
    int m = (int)strtol(gtk_entry_get_text((GtkEntry*)addressInput), NULL, 16);
    sprintf(tmpstr,"%02x %02x %02x %02x %02x %02x %02x %02x",
            memory[m   ],memory[m+ 1],memory[m+ 2],memory[m+ 3],
            memory[m+ 4],memory[m+ 5],memory[m+ 6],memory[m+ 7]);
    gtk_entry_set_text ((GtkEntry*)memCont1,tmpstr);
    sprintf(tmpstr,"%02x %02x %02x %02x %02x %02x %02x %02x",
            memory[m+ 8],memory[m+ 9],memory[m+10],memory[m+11],
            memory[m+12],memory[m+13],memory[m+14],memory[m+15]);
    gtk_entry_set_text ((GtkEntry*)memCont2,tmpstr);

}

gboolean on_idle( gpointer data );

void on_run(GtkWidget* widget,gpointer user_data)
{

    if (running==FALSE) {
        if (!context.halted) {
            running = TRUE;
            idle_id = gdk_threads_add_idle ( on_idle, NULL );
        }
    } else {
        running = FALSE;
        g_source_remove (idle_id);
    }

    if (running) {
        gtk_button_set_label ((GtkButton*)runButton,"Stop");
        gtk_widget_grab_focus (pluginPanel);
    } else {
        gtk_button_set_label ((GtkButton*)runButton,"Run");
    }
}


gboolean on_idle( gpointer data )
{
//    gdk_threads_enter();  // TODO ...and I'm supposed to do what instead
    if(!context.halted && running==TRUE) {
        Z80Execute(&context);
        //dump_z80_state();
    } else {
        running = TRUE;
        on_run(NULL, NULL);
    }


//    gdk_threads_leave();
    return( TRUE );
}

void on_step(GtkWidget* widget,gpointer user_data)
{
    if(!context.halted) {
        Z80Execute(&context);
        dump_z80_state();
    }
}


void loadBin(char *filename) {
	FILE *file = fopen( filename, "r" );
	if ( file == 0 ) {
		printf( "Could not open file\n" );
	} else {
		int x;
		long p=0;
		while  ( ( x = fgetc( file ) ) != EOF ) {
			memory[p++]=x;
		}
		fclose( file );
	}
}

void on_load(GtkWidget* widget,gpointer user_data)
{
    GtkWidget *dialog;
    GtkFileChooserAction action = GTK_FILE_CHOOSER_ACTION_OPEN;
    gint res;

    dialog = gtk_file_chooser_dialog_new ("Open File",
                                          GTK_WINDOW(mainwin),
                                          action,
                                          "_Cancel",
                                          GTK_RESPONSE_CANCEL,
                                          "_Open",
                                          GTK_RESPONSE_ACCEPT,
                                          NULL);

    gtk_file_chooser_set_current_folder ((GtkFileChooser*)dialog,"asm/");
    GtkFileFilter *filter = gtk_file_filter_new ();
    gtk_file_filter_add_pattern (filter, "*.bin");
    gtk_file_chooser_add_filter ((GtkFileChooser*)dialog, filter);

    res = gtk_dialog_run (GTK_DIALOG (dialog));
    if (res == GTK_RESPONSE_ACCEPT)
    {
        char *filename;
        GtkFileChooser *chooser = GTK_FILE_CHOOSER (dialog);
        filename = gtk_file_chooser_get_filename (chooser);

        loadBin(filename);

        g_free (filename);
    }

    gtk_widget_destroy (dialog);
}

static byte context_mem_read_callback(int param, ushort address)
{
    return memory[address];
}

static void context_mem_write_callback(int param, ushort address, byte data)
{
    
    memory[address] = data;

    for (GList* l = plugins; l != NULL; l = l->next) {
        plugInstStruct* i = (plugInstStruct*)l->data;
        int a = i->addressStart;
        int r = i->plug.getAddressSize();
        if (address>=a && address<(a+r)) {
            // address in range
            i->plug.setAddress(i, address, data);

        }
    }
}

static byte context_io_read_callback(int param, ushort address)
{
    int data=0xff;
    address = address & 0xff;
    gboolean handled = FALSE;
    for (GList* l = plugins; l != NULL; l = l->next) {
        plugInstStruct* i = (plugInstStruct*)l->data;
        int a = i->portStart;
        int r = i->plug.getPortSize();
        if (address>=a && address<a+r) {
            // IO port in range
            data = i->plug.getPort(i, address);
            handled = TRUE;
        }
    }
    if (!handled) {
        g_print("warning port write not handled by a plugin (0x%02x)=0x%02x\n",address,data);
    }

    return data;
}

static void context_io_write_callback(int param, ushort address, byte data)
{
    address = address & 0xff; // port address
    gboolean handled = FALSE;
    for (GList* l = plugins; l != NULL; l = l->next) {
        plugInstStruct* i = (plugInstStruct*)l->data;
        int a = i->portStart;
        int r = i->plug.getPortSize();
        if (address>=a && address<a+r) {
            // IO port in range
            i->plug.setPort(i, address, data);
            handled = TRUE;
        }
    }
    if (!handled) {
        g_print("warning port write not handled by a plugin (0x%02x)=0x%02x\n",address,data);
    }
}

void init_emulator()
{
    context.memRead = context_mem_read_callback;
    context.memWrite = context_mem_write_callback;
    context.ioRead = context_io_read_callback;
    context.ioWrite = context_io_write_callback;
}

static void setValText(GtkEntry* ent, ushort val)
{
    sprintf(tmpstr,"%04x",val);
    gtk_entry_set_text (ent,tmpstr);
}

static void dump_z80_state( void )
{
    /*
        printf( "AF %04x BC %04x DE %04x HL %04x\nAF'%04x BC'%04x DE'%04x HL'%04x IX%04x IY%04x SP%04x PC%04x\n",
                AF1, BC1, DE1, HL1, AF2, BC2, DE2, HL2, IX, IY, SP, PC );
        printf( "I%02x R%02x IFF1%d IFF2%d IM%d halted %d tstates %d\n", I, R,
                IFF1, IFF2, IM, context.halted, context.tstates );
    */
    setValText((GtkEntry*)AFtext, AF1);
    setValText((GtkEntry*)BCtext, BC1);
    setValText((GtkEntry*)DEtext, DE1);
    setValText((GtkEntry*)HLtext, HL1);

    setValText((GtkEntry*)IXtext, IX);
    setValText((GtkEntry*)IYtext, IY);
    setValText((GtkEntry*)SPtext, SP);
    setValText((GtkEntry*)PCtext, PC);

    on_addressChange(NULL,NULL);
}

// cache loaded plugin libraries
GHashTable* pluglookup;

// during xml attribute iteration each of these are filled in
// ready for the end of the element callback
char label[1024];
pluginStruct* plug;
int portStart,addressStart;

void startElement(void *data, const char *el, const char **attr) {
    int i;

    for (i = 0; attr[i]; i += 2) {
        if (g_strcmp0(attr[i],"libName")==0) {
            plug = (pluginStruct*)g_hash_table_lookup (pluglookup, attr[i + 1]);
            //g_print("\nplug %s - %lu\n",attr[i+1],plug);
            if (plug==NULL) {
                plug = malloc(sizeof(pluginStruct));
                sprintf(plug->libName,(const char*)attr[i + 1]);
                integratePlugin(plug);
                //g_print("integrated %s\n",attr[i + 1]);
                g_hash_table_insert (pluglookup, (gpointer)attr[i + 1],plug);
            }
        }
        if (g_strcmp0(attr[i],"label")==0) {
            sprintf(label,(const char*)attr[i + 1]);
        }
        if (g_strcmp0(attr[i],"portStart")==0) {
            portStart = (int)strtol((const char*)attr[i + 1], NULL, 16);
        }
        if (g_strcmp0(attr[i],"addressStart")==0) {
            addressStart = (int)strtol((const char*)attr[i + 1], NULL, 16);
        }
    }
}



void endElement(void *data, const char *el)
{
    if (plug!=NULL) {
        plugInstStruct*  pInst = malloc(sizeof(plugInstStruct));
        pInst->plug=*plug;
        namePluginInstance(pInst,label);
        setPluginInstanceStartPort(pInst,portStart);
        setPluginInstanceStartAddress(pInst,addressStart);
        pInst->plug.initialise(pInst,pluginPanel);

        plugins=g_list_append (plugins, pInst);
        sprintf(label,"\0");
        plug=NULL;
        portStart=0;
        addressStart=0;
    }

}


void centreWidget (GtkWidget *child, gpointer data) {
    // “halign”                   GtkAlign
    // GTK_ALIGN_CENTER
    gtk_widget_set_halign (child, GTK_ALIGN_CENTER);
}


int main( int argc, char **argv )
{
    for (int i=0; i<0x10000; i++) {
        memory[i] = 0;
    }
    init_emulator();

    GtkBuilder *gtkBuilder;

    gtk_init (&argc, &argv);

    gtkBuilder= gtk_builder_new();
    gtk_builder_add_from_file(gtkBuilder,"ui.glade",NULL);
    gtk_builder_connect_signals ( gtkBuilder, NULL );
    mainwin = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"mainWindow"));
    runButton = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"runButton"));

// is there an automagical way to do this...?
    AFtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"AFtext"));
    BCtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"BCtext"));
    DEtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"DEtext"));
    HLtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"HLtext"));

    IXtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"IXtext"));
    IYtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"IYtext"));
    SPtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"SPtext"));
    PCtext = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"PCtext"));

    addressInput = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"addressInput"));
    memCont1 = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"memCont1"));
    memCont2 = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"memCont2"));

    pluginPanel = GTK_WIDGET(gtk_builder_get_object(gtkBuilder,"pluginPanel"));
    gtk_widget_set_can_focus(pluginPanel,TRUE);

    
    g_object_unref ( G_OBJECT(gtkBuilder) );
    gtk_widget_show_all ( mainwin );


    pluglookup = g_hash_table_new ( g_str_hash , g_str_equal);

    g_print("requested machine config=%s\n",argv[1]);
    char buf[BUFSIZ];
    FILE *file = fopen(argv[1], "r" );
    if (file!=NULL) {
        XML_Parser parser = XML_ParserCreate(NULL);
        int done;
        int depth = 0;
        XML_SetUserData(parser, &depth);
        XML_SetElementHandler(parser, startElement, endElement);
        do {
            size_t len = fread(buf, 1, sizeof(buf), file);
            done = len < sizeof(buf);
            if (!XML_Parse(parser, buf, len, done)) {
                fprintf(stderr,
                        "%s at line %d\n",
                        XML_ErrorString(XML_GetErrorCode(parser)),
                        XML_GetCurrentLineNumber(parser));
                return 1;
            }
        } while (!done);
        XML_ParserFree(parser);
    }

    gtk_container_foreach (GTK_CONTAINER (pluginPanel), (GtkCallback) centreWidget, NULL);
    gtk_widget_show_all ( (GtkWidget*)mainwin );
    gtk_window_set_position(GTK_WINDOW(mainwin), GTK_WIN_POS_CENTER_ALWAYS);

	if (argc==3) {
		loadBin(argv[2]);
		printf("loaded %s\n",argv[2]);
	}
    
    gtk_main ();


    return 0;
}

